plugins/follow_me/follow_me.h

```diff
@@ -0,0 +1,102 @@
+#pragma once
+
+#include <iostream>
+#include <cmath>
+
+namespace dronecore {
+
+class FollowMeImpl;
+
+class FollowMe
+{
+public:
+    explicit FollowMe(FollowMeImpl *impl);
+    ~FollowMe();
+
+    /**
+     * @brief Results of FollowMe operations
+     */
+    enum class Result {
+        SUCCESS = 0, /**< @brief %Request succeeded. */
+        NO_DEVICE, /**< @brief No device connected. */
+        CONNECTION_ERROR, /**< @brief %Connection error. */
+        BUSY, /**< @brief Vehicle busy. */
+        COMMAND_DENIED, /**< @brief Command denied. */
+        TIMEOUT, /**< @brief %Request timeout. */
+        UNKNOWN /**< @brief Unknown error. */
+    };
+
+    /**
+     * @brief Altitude flag for Follow Target command.
+     */
+    enum class ESTCapabilities {
+        POS = 0,
+        VEL = 1,
+        ACCEL = 2,
+        ATT_RATES = 3
+    };
+
+    /**
+     * @brief Motion report info which will be emitted peridically to the Vehicle
```

Comment:
So what is the purpose of this? You're sending current motion of vehicle? Why does a user of the API need to know this/set this?

Reply:
@hamishwillee You're right. This should be auto-populated. But for testing purpose, we need to export Position (lat, long, alt) to API users. And, this structure will be part of impl and is auto-populated.

Reply:
@shakthi-prashanth-m (& @julianoes ) Just to be clear, are you saying that this will move to be internal?, and you'll test it via the internal API?
For me the API should look something like `start()`, `stop()`, `setFollowInfo(FollowMe::FollowInfo info)`, `getFollowInfo()`
Where `FollowInfo` would be info to specify the relative location that the follower should adopt. This might be a co-ordinate or "standard position + height" (front, back, front left, back left, height) or behaviour (e.g. circle me). 

I don't think anything else is needed, though possibly a string for printing output if we use standard positions (and hence have an enum)

Reply:
I agree with @hamishwillee. We don't need to internally check GPS. We just expose it via an API.

Reply:
As per https://docs.px4.io/en/flight_modes/follow_me.html QGroundControl doesn't prompt user for position. It captures current position & other GPS related details using platform-specific framework and sends to device. So, in our case we're setting position only for testing purpose.

Reply:
@shakthi-prashanth-m @julianoes 

> So, in our case we're setting position only for testing purpose.

No no no no. That is horrible. Nothing to do with testing should be in a public API file. Making it conditional does not make it any less ugly.  If you need to have some sort of test stuff, then have it in an implementation file and test that file. 

So I had another look at the mode docs. Updating my comment above in line with what PX4 does:

The API should be `start()`, `stop()`, `setFollowInfo(FollowMe::FollowInfo info)`, `getFollowInfo()`
Where `FollowInfo` would be and object that captures all the config params [here](https://docs.px4.io/en/flight_modes/follow_me.html#configuration).  You would then set/get the params as needed to drive the mode. You would have an enum and a string converter for `NAV_FT_FS`

Reply:
I agree about FollowInfo parameters. But somehow need to mock GPS position of the GCS and send periodic updates to device. It could be done internally, without a user API. As per FollowMe mode description [here](https://docs.px4.io/en/flight_modes/follow_me.html#configuration), GPS is needed. It's mentioned in the top,

**By using GPS and other positioning information a multicopter is able to automatically yaw to face and follow a user at a specified position and distance. While in this mode no user input is required.
This mode requires GPS.**

As discussed with @hamishwillee in slack, we can do as below:
1. Cleanup public header file `follow_me.h` and handle mocking position stuff in `follow_me_impl.h`.
2. Export public APIs for set/get FollowInfo
3. To use `CallEveryHandler` instead of `TimeoutHandler`, as per @julianoes suggestion.

Reply:
> No, why would you mock this position. Just feed it to DroneCore via the public API.

@julianoes Perhaps I'm dim, but what public API does DroneCore have for setting its **own**position? If such an API exists great - I wasn't aware we had platform-independent methods of getting position of dronecore in place.